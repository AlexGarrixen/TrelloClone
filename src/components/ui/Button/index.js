import React, { forwardRef } from 'react';
import PropTypes from 'prop-types';
import clsx from 'clsx';
import capitalize from '../../../utils/capitalize';
import classes from './classes';

const Button = forwardRef(
  (
    {
      children,
      className,
      component: Component = 'button',
      color = 'default',
      variant = 'contained',
      disabled = false,
      startIcon,
      endIcon,
      fullWidth = false,
      ...other
    },
    ref
  ) => (
    <Component
      {...other}
      ref={ref}
      disabled={disabled}
      className={clsx(
        classes.default,
        classes[`${variant}${capitalize(color)}`],
        disabled && classes.disabled,
        startIcon && classes.startIcon,
        endIcon && classes.endIcon,
        fullWidth && classes.fullWidth,
        className
      )}
    >
      {startIcon && startIcon}
      {children}
      {endIcon && endIcon}
    </Component>
  )
);

Button.propTypes = {
  children: PropTypes.node,
  component: PropTypes.elementType,
  color: PropTypes.oneOf(['default', 'primary', 'secondary']),
  variant: PropTypes.oneOf(['text', 'contained', 'outlined']),
  disabled: PropTypes.bool,
  startIcon: PropTypes.node,
  fullWidth: PropTypes.bool,
};

export default Button;
