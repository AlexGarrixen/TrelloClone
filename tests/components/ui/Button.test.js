import React from 'react';
import { mount } from 'enzyme';
import Button from '../../../src/components/ui/Button';

describe('<Button />', () => {
  test('Render success', () => {
    const wrapper = mount(<Button>Example</Button>);

    expect(wrapper).toMatchSnapshot();
  });

  test('Test with default props', () => {
    const wrapper = mount(<Button>Example</Button>);
    const element = wrapper.find('.button');

    expect(element.type()).toBe('button');
    expect(element.text()).toBe('Example');
    expect(element.hasClass('button--contained-default')).toBe(true);
  });

  test('Test with props variant and color', () => {
    const wrapper = mount(
      <Button variant='outlined' color='primary'>
        Example
      </Button>
    );
    const element = wrapper.find('.button');

    expect(element.text()).toBe('Example');
    expect(element.hasClass('button--outlined-primary')).toBe(true);
  });

  test('Test with prop startIcon', () => {
    const wrapper = mount(<Button startIcon={<svg />}>Example</Button>);
    const element = wrapper.find('.button');

    expect(element.find('svg').exists()).toBe(true);
  });

  test('Test with prop endIcon', () => {
    const wrapper = mount(<Button endIcon={<svg />}>Example</Button>);
    const element = wrapper.find('.button');

    expect(element.find('svg').exists()).toBe(true);
  });

  test('Test with prop fullWidth', () => {
    const wrapper = mount(<Button fullWidth>Example</Button>);
    const element = wrapper.find('.button');

    expect(element.hasClass('button--full-width')).toBe(true);
  });

  test('Test with prop component', () => {
    const wrapper = mount(<Button component='a'>Example</Button>);
    const element = wrapper.find('.button');

    expect(element.type()).toBe('a');
  });

  test('Test with prop disabled', () => {
    const wrapper = mount(<Button disabled>Example</Button>);
    const element = wrapper.find('.button');
    const props = element.props();

    expect(element.hasClass('disabled')).toBe(true);
    expect(props.disabled).toBe(true);
  });
});
