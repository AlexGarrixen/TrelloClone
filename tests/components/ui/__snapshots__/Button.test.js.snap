// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Button /> Render success 1`] = `
<button
  className="button button--contained-default"
  disabled={false}
  type="button"
>
  Example
</button>
`;
